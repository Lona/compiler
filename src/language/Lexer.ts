export type NextAction = { type: 'next'; value: string }

export type PushAction = { type: 'push'; value: string }

export type PopAction = { type: 'pop' }

export type Action = NextAction | PushAction | PopAction

export type Rule = {
  name: string
  pattern: string | RegExp
  action?: Action
}

export type StateDefinition = {
  name: string
  rules: Rule[]
}

export type SimpleStateDefinition = {
  name?: string
  rules: Omit<Rule, 'action'>[]
}

export type Token = {
  type: string
  value: string
  groups: string[]
  position: {
    start: number
    end: number
  }
}

export class Lexer {
  state: string[]
  stateDefinitions: StateDefinition[]

  constructor(
    states: SimpleStateDefinition | StateDefinition[],
    initialState?: string
  ) {
    const normalized: StateDefinition[] =
      states instanceof Array
        ? states
        : [{ ...states, name: states.name ?? 'main' }]
    this.state = [initialState ?? normalized[0].name]
    this.stateDefinitions = normalized
  }

  get currentState() {
    return this.state[this.state.length - 1]
  }

  private regExpCache: { [key: string]: RegExp } = {}

  getRegExp(pattern: string | RegExp) {
    const key = typeof pattern === 'string' ? pattern : pattern.source

    if (key in this.regExpCache) {
      return this.regExpCache[key]
    }

    const regExp = getRegExp(pattern)

    this.regExpCache[key] = regExp

    return regExp
  }

  tokenize(source: string) {
    let length = source.length
    let pos = 0
    let tokens: Token[] = []

    main: while (pos < length) {
      const currentState = this.currentState

      const stateDefinition = this.stateDefinitions.find(
        definition => definition.name === currentState
      )

      if (!stateDefinition) {
        throw new Error(`Invalid lexer state: ${currentState}`)
      }

      const { rules } = stateDefinition

      const slice = source.slice(pos)

      for (let rule of rules) {
        let regExp = this.getRegExp(rule.pattern)

        const match = slice.match(regExp)

        if (match) {
          const [value, ...groups] = match

          if (value.length === 0 && !rule.action) {
            throw new Error(
              'Lexer stalled: no input consumed and no action taken.'
            )
          }

          tokens.push({
            type: rule.name,
            value,
            groups,
            position: {
              start: pos,
              end: pos + value.length,
            },
          })

          if (rule.action) {
            switch (rule.action.type) {
              case 'next':
                this.state[this.state.length - 1] = rule.action.value
                break
              case 'push':
                this.state.push(rule.action.value)
                break
              case 'pop':
                this.state.pop
                break
            }
          }

          pos += value.length

          continue main
        }
      }

      console.error(
        `Failed to parse: ${slice.slice}\n`,
        currentState,
        rules.map(rule => rule.pattern)
      )

      break
    }

    return tokens
  }
}

function getRegExp(re: string | RegExp): RegExp {
  let source
  if (typeof re === 'string') {
    source = re
  } else if (re instanceof RegExp) {
    source = re.source
  } else {
    throw new Error(
      'rules must start with a match string or regular expression'
    )
  }

  const matchOnlyAtLineStart = source.length > 0 && source[0] === '^'

  return new RegExp(
    '^(?:' + (matchOnlyAtLineStart ? source.substr(1) : source) + ')'
  )
}
