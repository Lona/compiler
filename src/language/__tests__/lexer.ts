import { Lexer, Rule } from '../Lexer'

const keyword = (string: string): Rule => ({
  name: string,
  pattern: string,
  discard: false,
})

it('tokenizes simple tokens', () => {
  const lexer = new Lexer({
    rules: [
      keyword('from'),
      keyword('select'),
      keyword('where'),
      { name: 'identifier', pattern: '[a-zA-Z]\\w*' },
      { name: 'integer', pattern: '0|[1-9]\\d*' },
      { name: '_', pattern: '\\W+' },
    ],
  })

  const tokens = lexer.tokenize('select foo from 123')

  expect(tokens).toMatchSnapshot()
})

it('captures groups', () => {
  const lexer = new Lexer({
    rules: [
      {
        name: 'tag',
        pattern: '<(\\w+)>',
      },
    ],
  })

  const tokens = lexer.tokenize('<hi>')

  expect(tokens).toMatchSnapshot()
})

it('supports states', () => {
  const lexer = new Lexer([
    {
      name: 'main',
      rules: [
        {
          name: 'quote',
          pattern: '"',
          action: {
            type: 'next',
            value: 'string',
          },
        },
        {
          name: 'content',
          pattern: '[^"]+',
        },
      ],
    },
    {
      name: 'string',
      rules: [
        {
          name: 'quote',
          pattern: '"',
          action: {
            type: 'next',
            value: 'main',
          },
        },
        {
          name: 'string',
          pattern: '[^"]+',
        },
      ],
    },
  ])

  const tokens = lexer.tokenize('hello world "this is a string" ok')

  expect(tokens).toMatchSnapshot()
})

it('fails if no input is consumed', () => {
  const lexer = new Lexer({
    rules: [
      {
        name: 'nothing',
        pattern: '',
      },
    ],
  })

  expect(() => {
    lexer.tokenize('hello world "this is a string" ok')
  }).toThrow()
})
