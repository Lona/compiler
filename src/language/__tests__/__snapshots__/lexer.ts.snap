// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`captures groups 1`] = `
Array [
  Object {
    "groups": Array [
      "hi",
    ],
    "position": Object {
      "end": 4,
      "start": 0,
    },
    "type": "tag",
    "value": "<hi>",
  },
]
`;

exports[`supports states 1`] = `
Array [
  Object {
    "groups": Array [],
    "position": Object {
      "end": 12,
      "start": 0,
    },
    "type": "content",
    "value": "hello world ",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 13,
      "start": 12,
    },
    "type": "quote",
    "value": "\\"",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 29,
      "start": 13,
    },
    "type": "string",
    "value": "this is a string",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 30,
      "start": 29,
    },
    "type": "quote",
    "value": "\\"",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 33,
      "start": 30,
    },
    "type": "content",
    "value": " ok",
  },
]
`;

exports[`tokenizes simple tokens 1`] = `
Array [
  Object {
    "groups": Array [],
    "position": Object {
      "end": 6,
      "start": 0,
    },
    "type": "select",
    "value": "select",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 7,
      "start": 6,
    },
    "type": "_",
    "value": " ",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 10,
      "start": 7,
    },
    "type": "identifier",
    "value": "foo",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 11,
      "start": 10,
    },
    "type": "_",
    "value": " ",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 15,
      "start": 11,
    },
    "type": "from",
    "value": "from",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 16,
      "start": 15,
    },
    "type": "_",
    "value": " ",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 19,
      "start": 16,
    },
    "type": "integer",
    "value": "123",
  },
]
`;
